libyul/AST.h

```diff
@@ -36,38 +39,85 @@ namespace solidity::yul
 
 using Type = YulString;
 
-struct DebugData
+class DebugData
```

Comment:
This transition is not good :-). The ``struct`` was very simple before and designed towards staying immutable.
It staying immutable is important, since it's shared among multiple AST nodes as ``shared_ptr``s - modifying an existing ``DebugData`` will almost never have the effect you want.
So I'd roll this part back (the change in general mostly makes sense if you *want* mutable setters in here, which we don't)

Does that cause any issues elsewhere? I guess https://github.com/ethereum/solidity/pull/14841#discussion_r1484917216 was the main reason for this change in the first place, but you realized there yourself that it's better to use a new fresh copy of a newly constructed ``DebugData`` instead, right?

Reply:
Yeah, I think its ok to use only fresh copies of newly constructed `DebugData`. The current mechanism is allowing the creation of that using `constPtr()` that will create a new `shared_ptr` out of the data. Everywhere we expect `DebugData`, it always need to be a `shared_ptr<DebugData const>` (aka `DebugData::ConstPtr`).

In general I don't see a problem with having the oportunity t to easily create a copy, change attributes and finally create a `DebugData::ConstPtr` by using the `constPtr()` method.

Reply:
Ok, I just started to implement the parsing part and saw that this `constPtr()` mechanism is not helping in any way. So I just rolled back to the original simple `struct` and static `create` method mechanism.